Hsearch + quarkus blog post

yrodiere opened the pull request:
See https://staging.in.relation.to/2019/11/12/hibernate-search-quarkus/

Code available here: https://github.com/hibernate/hibernate-demos/pull/41

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+as native binaries with a tiny memory footprint and impressively fast boot time.
+
+The latter is what really sets Quarkus apart:
+it means that Quarkus-based applications are especially well-suited for running in a container.
```
At some point, we need to mention the live coding experience too, because that's the one thing that people really like and makes a real difference.

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+It integrates many Java libraries and provides tooling to ease development.
+Quarkus-based applications can run in a traditional HotSpot setup,
+but also (via https://www.graalvm.org/[GraalVM])
+as native binaries with a tiny memory footprint and impressively fast boot time.
```
I think I would separate the two sentences: it's important that people understand that Quarkus have outstanding benefits with HotSpot too.

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+
+[source]
+----
+$ mvn io.quarkus:quarkus-maven-plugin:0.28.0:create \
```
Let's use 1.0.0.CR1?

yrodiere replied:
I did, and updated all the examples. I just forgot to update this one.

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+
+=== Configuration properties
+
+The configuration of database access will go into Quarkus' main configuration file: `application.properties`:
```
Maybe for the first occurrence, you should git the full path?

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+%dev.POSTGRES_HOST=localhost <3>
+%dev.POSTGRES_DB=hsearch_demo
+%dev.POSTGRES_USER=hsearch_demo
+%dev.POSTGRES_PASSWORD=hsearch_demo
```
This is a bit weird.

yrodiere replied:
Ok, I'll need a hint. What is weird?

yrodiere replied:
As discussed, the fact that I override environment variables from this file. I'll change it.

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+[INFO] Changes detected - recompiling the module!
+[INFO] Compiling 9 source files to /home/yrodiere/workspaces/contributor-support/hibernate-demos/hibernate-search/hsearch-quarkus/target/classes
+[INFO]
+[INFO] --- quarkus-maven-plugin:1.0.0.CR1:dev (default-cli) @ hsearch-quarkus ---
```
See that's CR1 you're using!

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+2019-11-06 16:01:11,273 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (main) sequence "hibernate_sequence" does not exist, skipping
+2019-11-06 16:01:11,273 WARN  [io.agr.pool] (main) Datasource '<default>': JDBC resources leaked: 0 ResultSet(s) and 1 Statement(s)
+2019-11-06 16:01:11,346 WARN  [io.agr.pool] (main) Datasource '<default>': JDBC resources leaked: 0 ResultSet(s) and 1 Statement(s)
+2019-11-06 16:01:11,809 INFO  [io.quarkus] (main) Quarkus 1.0.0.CR1 started in 5.259s. Listening on: http://0.0.0.0:8080
```
5 seconds? This sounds like a lot?

yrodiere replied:
It does. Stopping and recompiling and running it again once or twice will bring down the startup time to 2.5s, so I suspect it has something to do with PostgreSQL being slow for the first few commands, but I don't know what is wrong exactly.

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+=== Mapping
+
+Hibernate Search is now aware of where to send indexed data,
+but it does not now what to send yet.
```
```suggestion
but it does not know what to send yet.
```

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+Hibernate Search is now aware of where to send indexed data,
+but it does not now what to send yet.
+
+The definition of which parts of the entities needs to be indexed in Elasticsearch is called the _mapping_.
```
```suggestion
The definition of which parts of the entities nees to be indexed in Elasticsearch is called the _mapping_.
```

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+<2> By default, the document sent to the index for each entity is empty, which is not very useful.
+New content is added by defining _fields_.
+Here we define a field whose content will be extracted from the `name` property.
+It is full-text field, i.e. a text field which will be split into words upon indexing.
```
```suggestion
It is a full-text field, i.e. a text field which will be split into words upon indexing.
```

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+
+We're now ready to start the application with Hibernate Search.
+
+Thanks to Quarkus' hot reload, if the application was already started with `quarkus:dev` when we performed the changes,
```
```suggestion
Thanks to Quarkus' live coding feature, if the application was already started with `quarkus:dev` when we performed the changes,
```

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+each time an entity is created, updated or deleted through a Hibernate ORM entity manager/session,
+Hibernate Search will detect these changes and reindex the relevant entities as appropriate.
+
+Automatic is enabled by default, is completely transparent and requires no configuration.
```
```suggestion
Automatic indexing is enabled by default, is completely transparent and requires no configuration.
```
?

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+<1> Declare a custom analyzer named `english`.
+<2> Set the tokenizer to `standard`, i.e. require that the analyzer generates words by splitting text on spaces, tabs, punctuation, etc.
+<3> Apply three token filters to transform the extracted words: `lowercase` which turns the words to lowercase,
+`stemmer_english` which is a custom filter (see below), `asciifolding` which replaces accented characters with they ascii counterpart
```
```suggestion
`stemmer_english` which is a custom filter (see below), `asciifolding` which replaces accented characters with their ascii counterpart
```

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+[source]
+----
+quarkus.hibernate-search.elasticsearch.analysis.configurer=org.hibernate.demos.quarkus.search.ClientElasticsearchAnalysisConfigurer
+----
```
See how cool it would be to just have an annotation ;)

yrodiere replied:
Yep, it's called `@Named` but for some obscure reason it won't work in the  Quarkus Hibernate Search extension :)

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+]
+----
+
+It worked: the test "laboratory" now matches the name "Aperture Science Laboratories".
```
```suggestion
It worked: the text "laboratory" now matches the name "Aperture Science Laboratories".
```
?

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+	@OneToMany(mappedBy = "assignedManager")
+	public List<Client> assignedClients = new ArrayList<>();
+
+	@FullTextField(analyzer = "english") // <1>
```
I would argue that this is not a good analyzer for names :).

yrodiere replied:
Fair enough, but I think after 1400 lines, I really should stop adding content to this post :) I'll add a comment about this near the end.

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+}
+----
+<1> Define the assigned manager as "indexed-embedded" into the client,
+meaning all the index fields defined in the business manager will be embedded into the client upon indexing.
```
```suggestion
meaning all the indexed fields defined in the business manager will be embedded into the client upon indexing.
```

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+----
+<1> Define the assigned manager as "indexed-embedded" into the client,
+meaning all the index fields defined in the business manager will be embedded into the client upon indexing.
+Simply put, a new field will appear in index documents generated for clients: `assignedManager.name`.
```
```suggestion
Simply put, a new field will appear in indexed documents generated for clients: `assignedManager.name`.
```
Not sure about that one, maybe you really want to talk about the index.

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+We now are the happy owners of a REST application
+providing both CRUD operations
+and more advanced full-text search operation,
+packaged as a container image.
```
You writing a haiku?

yrodiere replied:
A bad one, maybe :)

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+there are still things we could improve:
+
+* improving robustness by taking advantage of https://quarkus.io/guides/flyway[Flyway]
+to handle database schema uppgrade.
```
```suggestion
to handle database schema upgrade.
```

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+
+* improving robustness by taking advantage of https://quarkus.io/guides/flyway[Flyway]
+to handle database schema uppgrade.
+* adding https://quarkus.io/guides/native-and-ssl.html[support for SSL]
```
```suggestion
* adding https://quarkus.io/guides/native-and-ssl[support for SSL]
```

yrodiere replied:
Right. I got on this page from the website though, so you probably have an obsolete link somewhere.

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+including enums, numbers, date/time values, or even spatial coordinates (points).
+** even custom types can be indexed thanks to link:{hsearchDocUrl}#mapper-orm-bridge[custom bridges]
+** link:{hsearchDocUrl}#search-dsl-predicate[other predicates] are available,
+such as `range` ("between"/"greater than"/...), `within` (points within a circle) and many more.
```
I would talk about spatial predicates because it's not entirely obvious we have spatial.

gsmet commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+        -e POSTGRES_PASSWORD=hsearch_demo \
+        -e ES_HOSTS=http://localhost:9200 \
+        quarkus/hsearch-quarkus
+2019-11-07 10:33:15,520 INFO  [io.quarkus] (main) hsearch-quarkus 1.0-SNAPSHOT (running on Quarkus 1.0.0.CR1) started in 0.849s. Listening on: http://0.0.0.0:8080
```
This is with a native image? Looks weird that it doesn't start much faster. You are in prod mode right so you don't inject the data?

yrodiere replied:
I don't inject the data, but I initialize the schema. I believe that's why it's so slow. I guess I could write a few more scripts to initialize the DB and Elasticsearch, but I'm not sure it's reasonable to assume people will initialize the Elasticsearch schema by hand.

Strangely enough, the second run is a bit faster (0.5s) and the next ones much faster (0.1). Maybe it's related to Elasticsearch, I've noticed it takes a while to "warm up" and deliver decent response times.

yrodiere replied:
I added something about the warm-up.

gunnarmorling commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+
+=== What's Quarkus?
+
+https://quarkus.io/[Quarkus] is a Java stack for developing web applications.
```
It's more than web-apps, e.g. also stream processing or CLI tools eventually. It's a generic app dev stack.

yrodiere replied:
That's not what https://quarkus.io/ seems to focus on, but alright, I'll change it.

gunnarmorling commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+* https://www.docker.com/[Docker] and in particular https://docs.docker.com/compose/[Docker-compose]
+to run PostgreSQL and Elasticsearch.
+
+GraalVM is necessary in order to build a native image of the application;
```
It can be done via an in-container build, too, which I'd recommend for easier set-up.

yrodiere replied:
Right, I did that but forgot to update this line. Will do.

gunnarmorling commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+we told it we needed Hibernate Search.
+Thus the dependencies are already present in the POM.
+
+If you are curious, the dependency looks like this:
```
Perhaps show the add-extension stuff instead, for the better effect instead of having it in there from the beginning?

yrodiere replied:
Good idea! Will do.

gunnarmorling commented on posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc:
```diff
@@ -0,0 +1,1428 @@
+Successfully tagged quarkus/hsearch-quarkus:latest
+----
+
+The Docker image is now ready to be used.
```
container image

yrodiere commented:
Thanks to you both! I addressed your comments and force-pushed the result.

@Sanne , unless you have other comments, I will push this blog post soon.

yrodiere commented:
Discussed with Sanne, I'll release this on Tuesday.
